### DISCUSSION TITLE
Crazyflie building APPs, Params & custom APIs

### DISCUSSION THREAD

**Discussion by jonas-kupferschmid** [bitcraze/discussions#669] [Q&A]
I'm trying to create multiple youtube tutorials (from setup to programming projects) (to make Crazyflie more beginner-friendly). I wanted to make a small function to do a drone flip and an emergency stop with a button press on the laptop. However, I need some assistance and would be happy to receive help and any tips are appreciated!
I am having trouble finding out everything efficiently (can't find any good tutorials (one reason I am creating them) and hard to find my answers on documentation) so I thought I would come here for help. As most probably my questions are easy for you to answer. You don't have to answer them all but I really appreciate any help! 

Nr 1. Emergency stop 
I wanted to make an emergency stop when I press the Enter button on my laptop. I saw that in one of your tutorials "Crazyflie Logging and parameter framework tutorial" 8min 40s: https://www.youtube.com/watch?v=chWrNh73YBw&list=PLj9XMmQVSr-C-ZXeWWyZgeZ7C4g6A5nPj&index=9 
you guys did something similar. However, the video is 9 years old and I saw that most of it is completely different now. I saw that there is a variable emergencyStop now. Can I also add a param group with it and change it or is it smarter to create a new Parameter? And how can I let cfclient listen for key presses to enable emergency stop? (pressing the button or changing in parameters page is a bit slow)

Nr 2. Drone Flip 
I was able to perform a drone flip on a crazyflie for an ETH course, however there a custom firmware was used (for us to implement our own controller and functions). So I am having a hard time to reproduce it with the official firmware.
For the tutorial, I wanted to make a drone flip when pressing a button on the cfclient Flight control page. For that, I wanted to make an open loop flip, by manually setting the thrusts. What is the best way to have the program (potentially in the APP layer, if not doable, then in firmware is also fine) run at a certain frequency where I can set the thrusts of each motor without getting overwritten? Is it possible to write an API and how can I do that? 
Can I use the following existing params?
stabilizer.stop (is this enough, I saw that there are multiple controlers PID kalman etc, and default chooses 0 (auto), do i have to set one specific and turn off that specific controller?)
motorPowerSet.enable
motorPowerSet.m1-4 (what would the max value be that the motor can handle? when I set it)

Nr 3. Threading & ToF sensors
Do you have any tips on obstacle avoidance that I should pay attention to? Any general tips are appreciated! And as I will be using 4 ToF sensors each with 64px I thought it would be smart to multithread. Does Crazyflie support multithreading?

Nr 4. List of APIs
Is there a list of all the APIs that are supported and how they can be used in the APP layer? Or just some important ones for me to look out for, as https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/ isn't too well documented for me. 

This is the first time for me to work on a public GitHub so I need some more help, Thank you very much for helping me out!
Jonas Kupferschmid

**Comment by cc299792458** [comment]
Hey bro,

I'd like to anwser your 1st question. I know there is an api using crazyflie-lib-python to achieve emergencestop, you can check https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/commander/, the `def send_stop_setpoint(self)` function is to do that. 

I've also implement a code to use a butten to awake this function, just combined with tkinter. The role of tkinter is to wait for and recieve my input. When I press 'esc'(that's in my case, you can change to 'enter'), it will call `send_stop_setpoint` function and the crazyflie will stop suddenly.

**Reply by jonas-kupferschmid** [reply]
thank you for your reply, I'll look into it. Though I wanted to write it into the firmware and use cfclient for it. But maybe I'll also create a tutorial for the python client. ^^

**Comment by krichardsson** [comment]
Hi Jonas,
thanks for moving your questions to discussions! 
I feel a bit bad to ask you, but would it be possible for you to split your question up in multiple discussions? According to our "discussion rules" (#55) we want one topic per discussion to make it easy for other users to find information on specific topics. I should have mentioned this in the email, sorry!

I think your 4 points would make great discussion topics, except nr 3 that you could split up into a "threading" topic and an "obstacle avoidance" topic. 

Thanks!
Kristoffer

**Reply by jonas-kupferschmid** [reply]
ok thank you, I'll do that!

**Reply by krichardsson** [reply]
Thanks!
